[MLIR][RemoveDeadValues] Privatize public function with NonLive arguments before RDV.

navyxliu opened the pull request:
# Problem
Liveness analysis is inter-procedural. If there are some unused arguments in a public function, they propagate to callers. From the perspective of RemoveDeadValues, the signature of a public function is immutable. It can't cope with this situation. One side, it deletes outgoing arguments, on the other side it keeps the function intact.

# Solution
We deploy a pre-process pass called 'privatize-public-function'. It attempts to split interface and implementation if and only if 
1. liveness is inter-procedural
2. the target of callOp is a public function
3. there is at least one argument is NonLive.

We split interface and implementation. 
the original function is interface, the function body is just a forwarding call to the implementation. The new function is private as implementation. It derives the function body.

If IR has changed, we invalidate LivenessAnalysis and recompute for the following remove-dead-values. 


# Test plan
```
./bin/llvm-lit -v ../mlir/test/Transforms/remove-dead-values.mlir
before:

./input.mlir:13:5: error: null operand found
    call @immutable_fn_return_void_with_unused_argument(%arg0, %zero) : (i32, i32) -> ()
    ^
./input.mlir:13:5: note: see current operation: "func.call"(%arg0, <<NULL VALUE>>) <{callee = @immutable_fn_return_void_with_unused_argument}> : (i32, <<NULL TYPE>>) -> ()
after: pass
```

# Alternatives
1) createDummyArgument replaces arguments with dummies. eg. If the type can create a value with zero initializer, such as 'T x{0}'. The downside of this method is that we may not construct trivial values for all types, in particular the aggregate types.

2) We add another DenseSet called 'liveSet'. The initial values are the arguments of public functions. It propagates liveness backward just like Liveness analysis.

Commit:
[MLIR][RemoveDeadValues] Mark arguments of a public function Live

This diff also changes traversal order from forward to backward for
region/block/ops. This order guanratees Liveness updates at a callsite
can propagates to the defs of arguments.

```
./bin/llvm-lit -v ../mlir/test/Transforms/remove-dead-values.mlir
```

navyxliu committed:
Fix formatter issue.

navyxliu committed:
update processCallOp.

navyxliu committed:
Update the lit test and formatter.

navyxliu commented:
hi, @joker-eph 
I can't reopen PR160242 because I push my branch by force. Could you take a look this new patch?  thanks.

joker-eph commented on mlir/include/mlir/IR/Visitors.h:
```diff
@@ -39,6 +39,20 @@ struct ForwardIterator {
+      return llvm::reverse(range);
+    }
+  }
+};
```
> My intention is to visit basic block and CFG in reverse order.

Reverse order from what? The order of basic blocks isn't indicative of any execution order. The only constraint is on the first basic block to be the "entry" into the region.

navyxliu committed:
Add LIT test contains SCF

This diff also tries to propagate liveness recursively.

navyxliu committed:
[MLIR][RemoveDeadValues] Privatize public function beforehand.

navyxliu committed:
Update format.

navyxliu committed:
Update for formatter.  No trivial braces.

navyxliu committed:
fix a typo

joker-eph reviewed (commented):
I think this should be fine, I have some minor comments inline, PTAL

navyxliu committed:
Update per reviewer's feedbacks.

also return LogicalResult and exits early when it encounters an error.

navyxliu committed:
Merge branch 'llvm:main' into backward_remove_dead_values

joker-eph commented on mlir/lib/Transforms/RemoveDeadValues.cpp:
```diff
@@ -869,10 +870,151 @@ struct RemoveDeadValues : public impl::RemoveDeadValuesBase<RemoveDeadValues> {
+/// Returns true if any IR changes were made, false otherwise.
+static bool processCallOp(CallOpInterface callOp, ModuleOp moduleOp,
+                          RunLivenessAnalysis &la) {
+  Operation *callableOp = callOp.resolveCallable();
```
Yes I meant create a symbol table once for the duration of the pass and reuse it here to accelerate the queries.

navyxliu committed:
Use SymbolTable and also don't clone function body.

one optimization I made is that I don't clone function body. The function body
migrates to the new private function. After that, I create wrapper body for
the original function.
